### ISSUE TITLE
Decorated function must only accept positional arguments

### ISSUE THREAD

**Issue by oaishi** [pymanopt/pymanopt#140]
Hello, I am writing a solver for product manifold where my arguments of cost() function should be a list, instead of separate variables. I tried two variants of cost functions:

1. `def cost(X)`- which gave an error said: 

```
...\anaconda\Anaconda3_202105\lib\site-packages\pymanopt-0.2.5-py3.8.egg\pymanopt\autodiff\__init__.py in __call__(self, *args, **kwargs)
     50     def __call__(self, *args, **kwargs):
     51         assert self._compiled_function is not None
---> 52         return self._compiled_function(*args, **kwargs)
     53 
     54 

TypeError: cost() takes 1 positional argument but 2 were given
```

2. `def cost(*X)`:  which gave an error said:

```
...\anaconda\Anaconda3_202105\lib\site-packages\pymanopt-0.2.5-py3.8.egg\pymanopt\autodiff\__init__.py in decorator(*args, **kwargs)
     74             argspec = inspect.getfullargspec(function)
     75             if argspec.varargs or argspec.varkw or argspec.kwonlyargs:
---> 76                 raise ValueError(
     77                     "Decorated function must only accept positional "
     78                     "arguments")

ValueError: Decorated function must only accept positional arguments
```

Can you kindly suggest me how to design a desirable cost function that can work on list input?

**Comment by ninamiolane** [comment]
Interested in this too!

**Comment by qrebjock** [comment]
Have you tried with tuples? This is what is done for the `FixedRankEmbedded` class but it's not really a product manifold. This requires you need to define your own class to represent tangent vectors. I remember having an issue with Hessians when doing this though.

I am not sure it really helps because what you are trying to do is different. You could maybe share your code so that we can understand better what's going on.

**Comment by oaishi** [comment]
Hi, @qrebjock. Thanks for the reply! I tried with tuple, but unfortunately it does not work as well. I am sharing my code below, so that you can understand better:

```
# (1) Instantiate a manifold
manifold = Product([Stiefel(3, 3),] *  2)

# (2) Define the cost function (here using autograd.numpy)
@pymanopt.function.Autograd
def cost(X):     
    return np.sum(X[0]) + np.sum(X[1])

@pymanopt.function.Autograd
def cost_1(X, Y):     
    return np.sum(X) + np.sum(Y)

problem = Problem(manifold=manifold, cost=cost_1)
# problem = Problem(manifold=manifold, cost=cost)

# (3) Instantiate a Pymanopt solver
solver = SteepestDescent()

# let Pymanopt do the rest
Xopt, Yopt = solver.solve(problem, x = manifold.rand())
```

If you try running it, you will see cost_1() works correctly, but cost() doesn't. Both these functions were calculating the cost in the same way- except the cost() function takes a list argument. As we will be working on a dynamic number of manifolds, we needed to pass a list to the cost function instead of hardcoded arguments ( i.e: as in cost_1(X, Y) ).

I hope it helps you understand my problem.

**Comment by nkoep** [comment]
If you only are only interested about the product of Stiefel manifolds, then there's a workaround for that situation since we have special support for that in the Stiefel manifold class itself.

```python
import autograd.numpy as np

import pymanopt
from pymanopt import Problem
from pymanopt.manifolds import Product, Stiefel
from pymanopt.solvers import SteepestDescent

manifold = Stiefel(3, 3, k=2)


@pymanopt.function.Autograd
def cost(X):
    return np.sum(X[:, :, 0]) + np.sum(X[:, :, 1])


problem = Problem(manifold=manifold, cost=cost)

solver = SteepestDescent()

solution = solver.solve(problem, x=manifold.rand())
Xopt, Yopt = solution[:, :, 0], solution[:, :, 1]
```

**Comment by nkoep** [comment]
Closing this is currently solved in master (#155) and will be available in pymanopt 1.0.0.
